### PR TITLE
Definition

### PR DESCRIPTION
**Pull request by shashwatmishra76**
### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

### CODE REVIEW

File: coalib/bearlib/languages/definitions/KornShell.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+@Language
+class KornShell:
+    extensions = '.ksh'
+    comment_delimiters = '#',
+    multiline_comment_delimiters = {": '": "'"}
+    string_delimiters = {'"': '"', "'": "'", '`': '`'}
+    multiline_string_delimiters = string_delimiters
+    encapsulators = {'[': ']', '{': '}'}
+    keywords = [
+        'if', 'then', 'else', 'elif', 'fi', 'case', 'coproc', 'esac',
```
Comment (jayvdb) [inline coalib/bearlib/languages/definitions/KornShell.py]: Please check all keywords.
E.g. https://unix.stackexchange.com/questions/86270/how-do-you-use-the-command-coproc-in-various-shells

File: coalib/bearlib/languages/definitions/ZShell.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
```
Comment (jayvdb) [inline coalib/bearlib/languages/definitions/ZShell.py]: Please remove this file . And read your PR diff after uploading it to find such obvious mistakes before wasting reviewer time.

Comment (shashwatmishra76) [reply coalib/bearlib/languages/definitions/ZShell.py]: Correct me if I am wrong but I suppose I have added certain command names in ZShell file right? They too were reserved words and there was no concept of a floating point number in bash,so I added float to the zshell keywords.

File: coalib/bearlib/languages/definitions/Bash.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class Bash:
+    extensions = '.bash'
```
Comment (li-boxuan) [inline coalib/bearlib/languages/definitions/Bash.py]: This should be a tuple.

**Comment by shashwatmishra76** [comment]
@li-boxuan Can you review my changes once?

**Comment by shashwatmishra76** [comment]
@jayvdb I guess I have made the necessary changes.Please review.

File: tests/parsing/ConfParserTest.py.orig [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+import logging
+
+from coalib.parsing.ConfParser import ConfParser
+from coalib.settings.Section import Section
+
+
+class ConfParserTest(unittest.TestCase):
+    example_file = """setting = without_section
+    [foo]
+    to be ignored
```
Comment (pareksha) [inline tests/parsing/ConfParserTest.py.orig]: Your commit message says you are only adding language definitions, then why is this file here?

Comment (shashwatmishra76) [reply tests/parsing/ConfParserTest.py.orig]: I don't know how this got generated.Will remove this at once.

File: tests/TestUtilities.py [hunk 5]
```diff
@@ -61,6 +63,7 @@
     'Vala',
     'VisualBasic',
     'm4',
+    'ZShell',
```
Comment (jayvdb) [inline tests/TestUtilities.py]: this is not in alpha order.

**Comment by srivama** [comment]
Please refer to https://api.coala.io/en/latest/Developers/Writing_Good_Commits.html for writing commit messages.

File: tests/TestUtilities.py [hunk 6]
```diff
@@ -47,7 +47,6 @@
     'JavaScript Object Notation',
     'JavaServer Pages',
     'Jinja2',
-    'JSON',
```
Comment (jayvdb) [inline tests/TestUtilities.py]: why this change removing `JSON`?  How is this relevant to your change?

And please never rebase using a merge, like you did in https://github.com/coala/coala/pull/5712/commits/8a71a767691f8c .  It is ugly.  Now this PR has three commits.

Comment (shashwatmishra76) [reply tests/TestUtilities.py]: Because I was getting an assertion error in 
CollectorsTest.py : test_get_all_languages_without_unknown 
which said that 'JSON' is present in 'set(LANGUAGE_NAMES)' but not in 'languages '.

Comment (shashwatmishra76) [reply tests/TestUtilities.py]: Also,you already have 'JavaScript Object Notation' as a __qualname__ .So I don't think there is any need of having  JSON in TestUtilities.py when you already have 'JavaScript Object Notation' sitting there.

**Comment by shashwatmishra76** [comment]
@jayvdb Please review this.

File: tests/TestUtilities.py (hunk 6 above)
Comment (jayvdb) [reply tests/TestUtilities.py]: If you look at travis, the tests are passing without that change.... I see you have removed it now.  Thanks.

File: coalib/bearlib/languages/__init__.py [hunk 7]
```diff
@@ -16,8 +17,10 @@
 from .definitions.html import HTML
 from .definitions.Java import Java
 from .definitions.JavaScript import JavaScript
+from .definitions.Jinja2 import Jinja2
```
Comment (jayvdb) [inline coalib/bearlib/languages/__init__.py]: please do not re-arrange these.  That isnt what your issue is about.

Create a newcomer issue for someone else to sort them ;-)

File: tests/TestUtilities.py [hunk 8]
```diff
@@ -45,6 +46,8 @@
     'JavaScript',
     'JavaServer Pages',
     'Jinja2',
+    'KornShell',
+    'm4',
```
Comment (jayvdb) [inline tests/TestUtilities.py]: please dont move `m4`

File: coalib/bearlib/languages/definitions/Bash.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class Bash:
+    extensions = ('.bash')
```
Comment (ksdme) [inline coalib/bearlib/languages/definitions/Bash.py]: This needs to be a tuple.

File: coalib/bearlib/languages/definitions/Bash.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    comment_delimiters = '#',
+    multiline_comment_delimiters = {": '": "'"}
+    string_delimiters = {'"': '"', "'": "'", '`': '`'}
+    multiline_string_delimiters = string_delimiters
+    encapsulators = {'[': ']', '{': '}'}
+    keywords = [
+        'if', 'then', 'else', 'elif', 'fi', 'case', 'esac',
+        'for', 'select', 'while', 'until', 'do', 'done', 'in',
+        'function', 'time']
```
Comment (ksdme) [inline coalib/bearlib/languages/definitions/Bash.py]: Trailing comma.

File: coalib/bearlib/languages/definitions/KornShell.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class KornShell:
+    extensions = ('.ksh')
```
Comment (ksdme) [inline coalib/bearlib/languages/definitions/KornShell.py]: Needs to be a tuple.

File: coalib/bearlib/languages/definitions/KornShell.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    comment_delimiters = '#',
+    multiline_comment_delimiters = {": '": "'"}
+    string_delimiters = {'"': '"', "'": "'", '`': '`'}
+    multiline_string_delimiters = string_delimiters
+    encapsulators = {'[': ']', '{': '}'}
+    keywords = [
+        'if', 'then', 'else', 'elif', 'fi', 'case', 'esac',
+        'for', 'select', 'while', 'until', 'do', 'done', 'in',
+        'function', 'time']
```
Comment (ksdme) [inline coalib/bearlib/languages/definitions/KornShell.py]: Trailing comma.

File: coalib/bearlib/languages/definitions/Shell.py [hunk 13]
```diff
@@ -3,9 +3,7 @@
 
 @Language
 class Shell:
-    aliases = 'bash', 'sh'
-
-    extensions = '.sh', '.bash', '.zsh'
+    extensions = ('.sh')
```
Comment (ksdme) [inline coalib/bearlib/languages/definitions/Shell.py]: Tuple.

File: coalib/bearlib/languages/definitions/Shell.py [hunk 14]
```diff
@@ -15,4 +13,4 @@ class Shell:
         'if', 'then', 'else', 'elif', 'fi', 'case', 'coproc', 'esac',
         'for', 'select', 'while', 'until', 'do', 'done', 'in',
         'function']
```
Comment (ksdme) [inline coalib/bearlib/languages/definitions/Shell.py]: Comma.

File: coalib/bearlib/languages/definitions/ZShell.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+class ZShell:
+    extensions = ('.zsh')
```
Comment (ksdme) [inline coalib/bearlib/languages/definitions/ZShell.py]: Tuple.

File: coalib/bearlib/languages/definitions/ZShell.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+    string_delimiters = {'"': '"', "'": "'", '`': '`'}
+    multiline_string_delimiters = string_delimiters
+    encapsulators = {'[': ']', '{': '}'}
+    keywords = [
+        'do', 'done', 'esac', 'then', 'elif', 'else',
+        'fi', 'for', 'case', 'if', 'while', 'function',
+        'repeat', 'time', 'until', 'select', 'coproc',
+        'nocorrect', 'foreach', 'end', 'declare',
+        'export', 'float', 'integer', 'local',
+        'readonly', 'typeset']
```
Comment (ksdme) [inline coalib/bearlib/languages/definitions/ZShell.py]: Comma.

**Comment by ksdme** [comment]
LGTM except for https://github.com/coala/coala/pull/5712#discussion_r214536809.

File: coalib/bearlib/languages/definitions/Shell.py [hunk 17]
```diff
@@ -14,5 +12,5 @@ class Shell:
     keywords = [
         'if', 'then', 'else', 'elif', 'fi', 'case', 'coproc', 'esac',
         'for', 'select', 'while', 'until', 'do', 'done', 'in',
-        'function']
-    special_chars = list('{}!|><[]')
+        'function', ]
```
Comment (jayvdb) [inline coalib/bearlib/languages/definitions/Shell.py]: do not modify this line.

File: coalib/bearlib/languages/definitions/ZShell.py [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+    string_delimiters = {'"': '"', "'": "'", '`': '`'}
+    multiline_string_delimiters = string_delimiters
+    encapsulators = {'[': ']', '{': '}'}
+    keywords = [
+        'do', 'done', 'esac', 'then', 'elif', 'else',
+        'fi', 'for', 'case', 'if', 'while', 'function',
+        'repeat', 'time', 'until', 'select', 'coproc',
+        'nocorrect', 'foreach', 'end', 'declare',
+        'export', 'float', 'integer', 'local',
+        'readonly', 'typeset', ]
```
Comment (jayvdb) [inline coalib/bearlib/languages/definitions/ZShell.py]: style "trailing comma" is useless unless the comma is at the end of the line.

File: coalib/bearlib/languages/definitions/KornShell.py [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+    extensions = ('.ksh',)
+    comment_delimiters = '#',
+    multiline_comment_delimiters = {": '": "'"}
+    string_delimiters = {'"': '"', "'": "'", '`': '`'}
+    multiline_string_delimiters = string_delimiters
+    encapsulators = {'[': ']', '{': '}'}
+    keywords = [
+        'if', 'then', 'else', 'elif', 'fi', 'case', 'esac',
+        'for', 'select', 'while', 'until', 'do', 'done', 'in',
+        'function', 'time', ]
```
Comment (jayvdb) [inline coalib/bearlib/languages/definitions/KornShell.py]: trailing comma is useless unless the comma is at the end of the line.

File: coalib/bearlib/languages/definitions/ZShell.py [hunk 20]
```diff
@@ -0,0 +1,19 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class ZShell:
+    extensions = ('.zsh',)
+    comment_delimiters = '#',
```
Comment (jayvdb) [inline coalib/bearlib/languages/definitions/ZShell.py]: on the line above, you use round brackets.  here you dont.  self-consistency is important.

check what style flake8 plugin and pypi package `hacking` uses for tuples?

there is also `flake8-comprehensions` which might have a preference.

(we dont use flake8, or its many useful plugins, but we should..)

File: coalib/bearlib/languages/definitions/Bash.py (hunk 10 above)
Comment (jayvdb) [reply coalib/bearlib/languages/definitions/Bash.py]: @shashwatmishra76 see https://duckduckgo.com/?q=trailing+commas+python&t=qupzilla&ia=qa

File: coalib/bearlib/languages/definitions/Shell.py [hunk 21]
```diff
@@ -14,5 +12,6 @@ class Shell:
     keywords = [
         'if', 'then', 'else', 'elif', 'fi', 'case', 'coproc', 'esac',
         'for', 'select', 'while', 'until', 'do', 'done', 'in',
-        'function']
-    special_chars = list('{}!|><[]')
+        'function',
```
Comment (jayvdb) [inline coalib/bearlib/languages/definitions/Shell.py]: do not modify this line.

when you write new lines, or modify existing lines, do you very best.

when you dont need to touch other lines, do not touch them.

**Comment by ksdme** [comment]
@shashwatmishra76 Please rebase this repo.

**Comment by ksdme** [comment]
@jayvdb PTAL.

**Comment by jayvdb** [comment]
ack https://github.com/coala/coala/commit/d5c0b794c85fe602b397ad730da983fb4a3428eb

**Comment by ksdme** [comment]
@jayvdb PTALA.

**Comment by jayvdb** [comment]
@gitmate-bot ff
